src/main/host/syscall/handler/mman.rs

```diff
@@ -109,13 +117,325 @@ impl SyscallHandler {
+                // `_tcp_endOfFileSignalled`), so we can't make this a panic.
+                log::warn!("File {file:p} (fd={fd}) is closed");
+                return Err(Errno::EBADF.into());
+            }
+
+            if unsafe { c::legacyfile_getType(file) } != c::_LegacyFileType_DT_FILE {
+                log::debug!("Descriptor exists for fd {fd}, but is not a regular file type");
+                return Err(Errno::EACCES.into());
+            }
+
+            // success; we know we have a file type descriptor
+            Some(file as *mut c::RegularFile)
+        };
+
+        // this fd exists in the plugin and not shadow
+        let plugin_fd = file.map(|file| Self::open_plugin_file(ctx.objs, fd, file));
+
+        // the file is None for an anonymous mapping, or a non-null Some otherwise
+        let Ok(plugin_fd) = plugin_fd.transpose() else {
+            log::warn!("mmap on fd {fd} for {len} bytes failed");
+            return Err(Errno::EACCES.into());
+        };
+
+        let flags = flags.bits();
+
+        // check for truncated flag bits (use u32 instead of i32 to prevent sign extension when
+        // casting from signed to unsigned)
+        if flags as u32 as u64 != flags {
+            warn_once_then_trace!("(LOG_ONCE) ignoring truncated flags from mmap: {flags}");
+        }
+
+        let flags = flags as i32;
+
+        // delegate execution of the mmap itself to the memory manager
+        let mut memory_manager = ctx.objs.process.memory_borrow_mut();
+        let mmap_result = memory_manager.do_mmap(
+            ctx.objs,
+            addr,
+            len,
+            prot,
+            flags,
+            plugin_fd.unwrap_or(-1),
+            offset,
+        );
+
+        if matches!(mmap_result, Err(SyscallError::Native)) {
+            let (ctx, thread) = ctx.objs.split_thread();
+            return Ok(thread
+                .native_mmap(
+                    &ctx,
+                    addr,
+                    len,
+                    prot,
+                    flags,
+                    plugin_fd.unwrap_or(-1),
+                    offset,
+                )?
+                .into());
+        }
+
+        log::trace!(
+            "Plugin-native mmap syscall at plugin addr {addr:p} with plugin fd {fd} for \
+            {len} bytes returned {mmap_result:?}"
+        );
+
+        // close the file we asked them to open
+        if let Some(plugin_fd) = plugin_fd {
+            Self::close_plugin_file(ctx.objs, plugin_fd);
+        }
+
+        mmap_result
+    }
+
+    fn open_plugin_file(
+        ctx: &mut ThreadContext,
+        fd: std::ffi::c_ulong,
+        file: *mut c::RegularFile,
+    ) -> Result<i32, ()> {
+        assert!(!file.is_null());
+
+        log::trace!("Trying to open file {fd} in the plugin");
+
+        // TODO: make sure we don't open special files like /dev/urandom, /etc/localtime etc. in the
+        // plugin via mmap
```

Comment:
Would it be easy enough to at least detect when this is happening and log a warning?

Reply:
I added a check for the few cases that shadow supports.